rust/lance/src/io/exec/filtered_read.rs

```diff
@@ -520,158 +470,291 @@ impl FilteredReadStream {
-                range_offset += num_physical_rows;
-                range_start + num_physical_rows
-            } else {
-                range_offset += num_logical_rows;
-                range_start + num_logical_rows
-            };
+                row_id_sequence,
+                to_read,
+                &mut to_skip,
+                &mut to_take,
+                &mut fragments_to_read,
+                &mut scan_push_down_fragments_to_read,
+            );
 
-            // By default we assume we will need to apply the full filter
-            // This will get refined if we have an exact match
-            let mut filter = &full_filter;
-
-            let mut to_read: Vec<Range<u64>> = if let Some(evaluated_index) = evaluated_index {
-                if evaluated_index
-                    .applicable_fragments
-                    .contains(fragment.id() as u32)
-                {
-                    let _span =
-                        tracing::span!(tracing::Level::DEBUG, "apply_index_result").entered();
-                    // There is an index result, and it applies to this fragment, so we can maybe
-                    // reduce the amount of data we read, and we can potentially push the scan range
-                    // down
-                    match &evaluated_index.index_result {
-                        IndexExprResult::Exact(row_id_mask) => {
-                            // The index result is an exact match, so we can both reduce the amount
-                            // of data we read, and we can push a filtered scan range down
-
-                            // Also, with an exact match, we only need to apply the refine filter
-                            filter = &refine_filter;
-
-                            let valid_ranges = row_id_sequence.mask_to_offset_ranges(row_id_mask);
-                            if let Some(deletion_vector) = &deletion_vector {
-                                Self::filter_deleted_rows(valid_ranges, deletion_vector)
-                            } else {
-                                valid_ranges
-                            }
-                        }
-                        IndexExprResult::AtMost(row_id_mask) => {
-                            // The index result is an at most, so we can reduce the amount of data
-                            // we read, but we can't push a filtered scan range down
-                            let valid_ranges = row_id_sequence.mask_to_offset_ranges(row_id_mask);
-                            if let Some(deletion_vector) = &deletion_vector {
-                                Self::filter_deleted_rows(valid_ranges, deletion_vector)
-                            } else {
-                                valid_ranges
+            if to_take == 0 {
+                scan_planned_with_limit_pushed_down = true;
+                for (frag_id, ranges) in scan_push_down_fragments_to_read {
+                    fragments_to_read.insert(frag_id, ranges);
```

Comment:
Why do we need to insert here?  Don't we always insert in `fragments_to_read` whenever we insert into `scan_push_down_fragments_to_read`?

Reply:
when to_take is zero, we should use the fragments in `scan_push_down_fragments_to_read` only.